File: data_structures/nft.py
class NFT:
    """Simple non-fungible token."""

    def __init__(self, nft_id, creator, owner, metadata, price=0.0, listed=False):
        self.id = nft_id
        self.creator = creator
        self.owner = owner
        self.metadata = metadata
        self.price = price
        self.listed = listed


class NFTMarketplace:
    """Minimal marketplace to trade :class:`NFT` objects."""

    def __init__(self, event_bus=None):
        self.event_bus = event_bus
        self.nfts = []
        self._next_id = 0

    def mint_nft(self, creator, metadata, price=0.0, listed=True):
        nft = NFT(self._next_id, creator, creator, metadata, price, listed)
        self.nfts.append(nft)
        self._next_id += 1
        step = getattr(creator.model, "current_step", 0)
        if self.event_bus:
            self.event_bus.publish(
                "nft_minted",
                step=step,
                nft_id=nft.id,
                creator=creator.unique_id,
                price=price,
            )
        if listed:
            self.list_nft(nft.id, price, step=step)
        return nft

    def list_nft(self, nft_id, price=None, *, step=0):
        nft = self._get_nft(nft_id)
        if nft is None:
            return False
        if price is not None:
            nft.price = price
        nft.listed = True
        if self.event_bus:
            self.event_bus.publish(
                "nft_listed",
                step=step,
                nft_id=nft.id,
                price=nft.price,
                owner=getattr(nft.owner, "unique_id", None),
            )
        return True

    def buy_nft(self, buyer, nft_id):
        nft = self._get_nft(nft_id)
        if nft is None or not nft.listed or buyer.tokens < nft.price:
            return False
        seller = nft.owner
        buyer.tokens -= nft.price
        seller.tokens += nft.price
        nft.owner = buyer
        nft.listed = False
        step = getattr(buyer.model, "current_step", 0)
        if self.event_bus:
            self.event_bus.publish(
                "nft_sold",
                step=step,
                nft_id=nft.id,
                seller=getattr(seller, "unique_id", None),
                buyer=buyer.unique_id,
                price=nft.price,
            )
        return True

    def transfer_nft(self, nft_id, new_owner):
        nft = self._get_nft(nft_id)
        if nft is None:
            return False
        old_owner = nft.owner
        nft.owner = new_owner
        nft.listed = False
        step = getattr(new_owner.model, "current_step", 0)
        if self.event_bus:
            self.event_bus.publish(
                "nft_transferred",
                step=step,
                nft_id=nft.id,
                old_owner=getattr(old_owner, "unique_id", None),
                new_owner=new_owner.unique_id,
            )
        return True

    def get_listed_nfts(self):
        return [n for n in self.nfts if n.listed]

    def _get_nft(self, nft_id):
        for nft in self.nfts:
            if nft.id == nft_id:
                return nft
        return None

File: lib/data-structures/nft.ts
// NFT and NFT Marketplace

import type { EventBus } from '../utils/event-bus';
import type { DAOMember } from '../agents/base';

export class NFT {
  id: number;
  creator: string; // Creator's unique ID
  owner: string; // Owner's unique ID
  metadata: any;
  price: number;
  listed: boolean;

  constructor(
    id: number,
    creator: string,
    owner: string,
    metadata: any,
    price: number = 0,
    listed: boolean = false
  ) {
    this.id = id;
    this.creator = creator;
    this.owner = owner;
    this.metadata = metadata;
    this.price = price;
    this.listed = listed;
  }
}

export class NFTMarketplace {
  eventBus: EventBus | null;
  nfts: NFT[] = [];
  private nextId: number = 0;

  constructor(eventBus: EventBus | null = null) {
    this.eventBus = eventBus;
  }

  mintNFT(creator: DAOMember, metadata: any, price: number = 0, listed: boolean = true): NFT {
    const nft = new NFT(this.nextId, creator.uniqueId, creator.uniqueId, metadata, price, listed);
    this.nfts.push(nft);
    this.nextId++;

    const step = creator.model?.currentStep || 0;

    if (this.eventBus) {
      this.eventBus.publish('nft_minted', {
        step,
        nftId: nft.id,
        creator: creator.uniqueId,
        price,
      });
    }

    if (listed) {
      this.listNFT(nft.id, price, step);
    }

    return nft;
  }

  listNFT(nftId: number, price?: number, step: number = 0): boolean {
    const nft = this.getNFT(nftId);
    if (!nft) return false;

    if (price !== undefined) {
      nft.price = price;
    }
    nft.listed = true;

    if (this.eventBus) {
      this.eventBus.publish('nft_listed', {
        step,
        nftId: nft.id,
        price: nft.price,
        owner: nft.owner,
      });
    }

    return true;
  }

  buyNFT(buyer: DAOMember, nftId: number): boolean {
    const nft = this.getNFT(nftId);
    if (!nft || !nft.listed || buyer.tokens < nft.price) {
      return false;
    }

    // Find seller (owner) - we need to get the actual member object
    // This assumes we have access to the model with members
    const seller = buyer.model?.dao?.members.find(m => m.uniqueId === nft.owner);
    if (!seller) return false;

    // Transfer tokens
    buyer.tokens -= nft.price;
    seller.tokens += nft.price;

    // Transfer ownership
    nft.owner = buyer.uniqueId;
    nft.listed = false;

    const step = buyer.model?.currentStep || 0;

    if (this.eventBus) {
      this.eventBus.publish('nft_sold', {
        step,
        nftId: nft.id,
        seller: seller.uniqueId,
        buyer: buyer.uniqueId,
        price: nft.price,
      });
    }

    return true;
  }

  transferNFT(nftId: number, newOwner: DAOMember): boolean {
    const nft = this.getNFT(nftId);
    if (!nft) return false;

    const oldOwner = nft.owner;
    nft.owner = newOwner.uniqueId;
    nft.listed = false;

    const step = newOwner.model?.currentStep || 0;

    if (this.eventBus) {
      this.eventBus.publish('nft_transferred', {
        step,
        nftId: nft.id,
        oldOwner,
        newOwner: newOwner.uniqueId,
      });
    }

    return true;
  }

  getListedNFTs(): NFT[] {
    return this.nfts.filter(n => n.listed);
  }

  getNFT(nftId: number): NFT | null {
    return this.nfts.find(n => n.id === nftId) || null;
  }
}
